include/machine/io.h

```diff
@@ -7,33 +7,109 @@
 #pragma once
 
 #include <config.h>
-#include <util.h>
+
+#if defined(CONFIG_PRINTING) || defined(CONFIG_DEBUG_BUILD)
+
 #include <arch/types.h>
+#include <stdarg.h>
 
-#define FORMAT(archetype, string_index, first_to_check) \
-        __attribute__((format(archetype, string_index, first_to_check)))
 
-#if (defined CONFIG_DEBUG_BUILD) || (defined CONFIG_PRINTING)
-void putDebugChar(unsigned char c);
+#ifdef CONFIG_PRINTING
+void kernel_putDebugChar(unsigned char c);
 #endif
 
 #ifdef CONFIG_DEBUG_BUILD
 /* io for dumping capdl */
-unsigned char getDebugChar(void);
+unsigned char kernel_getDebugChar(void);
 #endif
 
+
 #ifdef CONFIG_PRINTING
-/* printf will result in output */
-void putchar(char c);
-#define kernel_putchar(c) putchar(c)
-word_t kprintf(const char *format, ...) VISIBLE FORMAT(printf, 1, 2);
-word_t ksnprintf(char *str, word_t size, const char *format, ...);
-word_t puts(const char *s) VISIBLE;
-#define printf(args...) kprintf(args)
-#define snprintf(args...) ksnprintf(args)
-#else /* CONFIG_PRINTING */
-/* printf will NOT result in output */
-#define kernel_putchar(c) ((void)(0))
-#define printf(args...) ((void)(0))
-#define puts(s) ((void)(0))
+
+/* this is the actual implementation of the kernel printing API */
+void impl_kvprintf(const char *format, va_list ap);
```

Comment:
Am I right in concluding that some changes to the printf layer output channel management are logically different from the changes inserting a console layer to the output drivers?  Are you able to split these into two commits to make it clearer to see the out_wrap_t changes in isolation?